relay-metrics/src/aggregator.rs

```diff
@@ -510,6 +474,56 @@ impl fmt::Debug for CostTracker {
     }
 }
 
+/// Returns the instant at which a bucket should be flushed.
+///
+/// Recent buckets are flushed after a grace period of `initial_delay`. Backdated buckets, that
+/// is, buckets that lie in the past, are flushed after the shorter `debounce_delay`.
+fn get_flush_time(
+    config: &AggregatorConfig,
+    base_instant: Instant,
+    bucket_key: &BucketKey,
+) -> Instant {
+    let initial_flush = bucket_key.timestamp + config.bucket_interval() + config.initial_delay();
+
+    let now = UnixTimestamp::now();
+    let backdated = initial_flush <= now;
+
+    let delay = now.as_secs() as i64 - bucket_key.timestamp.as_secs() as i64;
+    relay_statsd::metric!(
+        histogram(MetricHistograms::BucketsDelay) = delay as f64,
+        backdated = if backdated { "true" } else { "false" },
+    );
+
+    let flush_timestamp = if backdated {
+        // If the initial flush time has passed or can't be represented, we want to treat the
+        // flush of the bucket as if it came in with the timestamp of current bucket based on
+        // the now timestamp.
+        //
+        // The rationale behind this is that we want to flush this bucket in the earliest slot
+        // together with buckets that have similar characteristics (e.g., same partition,
+        // project...).
+        let floored_timestamp = (now.as_secs() / config.bucket_interval) * config.bucket_interval;
+        UnixTimestamp::from_secs(floored_timestamp)
+            + config.bucket_interval()
+            + config.debounce_delay()
+    } else {
+        // If the initial flush is still pending, use that.
+        initial_flush
+    };
+
+    let instant = if flush_timestamp > now {
+        Instant::now().checked_add(flush_timestamp - now)
+    } else {
+        Instant::now().checked_sub(now - flush_timestamp)
+    }
+    .unwrap_or_else(Instant::now);
+
+    // Since `Instant` doesn't allow to get directly how many seconds elapsed, we leverage the
+    // diffing to get a duration and round it to the smallest second.
+    let instant = base_instant + Duration::from_secs((instant - base_instant).as_secs());
+    instant + config.flush_time_shift(bucket_key)
+}
```

Comment:
Is there a good reason for moving this function? It makes it harder to spot what has changed.

Reply:
With the introduction of the reference time, I don't consider this 'config' anymore (was questionable before). I suggested to move it and make it a private detail of the aggregator.

Reply:
@Dav1dde suggested moving it out of the config because he didn't want the config as self and the base instant supplied inside. He might motivate his choice better.

Reply:
I missed that it's a standalone function now. Thank you, all clear!